File: frontend/src/api/teamMembers.ts
export type TeamMember = {
  id: number;
  name: string;
  role: string;
  department: string;
  is_active: boolean;
  created_at: string;
};

export type TeamMemberPayload = {
  name: string;
  role: string;
  department: string;
};

async function readJson<T>(response: Response): Promise<T> {
  if (!response.ok) {
    const body = await response.json().catch(() => ({ detail: "요청 처리 중 오류가 발생했습니다." }));
    throw new Error(body.detail ?? "요청 처리 중 오류가 발생했습니다.");
  }

  return response.json();
}

export async function fetchTeamMembers(): Promise<TeamMember[]> {
  const data = await readJson<{ items: TeamMember[] }>(await fetch("/api/team-members"));
  return data.items;
}

export async function createTeamMember(payload: TeamMemberPayload): Promise<TeamMember> {
  return readJson<TeamMember>(
    await fetch("/api/team-members", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }),
  );
}

export async function updateTeamMember(id: number, payload: TeamMemberPayload): Promise<TeamMember> {
  return readJson<TeamMember>(
    await fetch(`/api/team-members/${id}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }),
  );
}

export async function deleteTeamMember(id: number): Promise<void> {
  const response = await fetch(`/api/team-members/${id}`, { method: "DELETE" });
  if (!response.ok) {
    const body = await response.json().catch(() => ({ detail: "팀원 삭제 중 오류가 발생했습니다." }));
    throw new Error(body.detail ?? "팀원 삭제 중 오류가 발생했습니다.");
  }
}

File: backend/app/routers/team_members.py
from typing import Annotated

from fastapi import APIRouter, HTTPException, Query, status
from pydantic import BaseModel

from app.services import team_member_service

router = APIRouter(prefix="/team-members", tags=["team-members"])


class TeamMemberPayload(BaseModel):
    name: str
    role: str = ""
    department: str = ""


@router.get("")
def list_team_members(
    include_inactive: Annotated[bool, Query(alias="include_inactive")] = False,
) -> dict[str, list[dict[str, object]]]:
    return {"items": team_member_service.list_team_members(include_inactive)}


@router.post("", status_code=status.HTTP_201_CREATED)
def create_team_member(payload: TeamMemberPayload) -> dict[str, object]:
    try:
        return team_member_service.create_team_member(
            name=payload.name,
            role=payload.role,
            department=payload.department,
        )
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.put("/{member_id}")
def update_team_member(member_id: int, payload: TeamMemberPayload) -> dict[str, object]:
    try:
        return team_member_service.update_team_member(
            member_id=member_id,
            name=payload.name,
            role=payload.role,
            department=payload.department,
        )
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    except LookupError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc


@router.delete("/{member_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_team_member(member_id: int) -> None:
    try:
        team_member_service.deactivate_team_member(member_id)
    except LookupError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
